api/observation.go

```diff
@@ -0,0 +1,389 @@
+		// Ignore case sensitivity
+		dimension := strings.ToLower(rawDimension)
+
+		queryParamExists := false
+		for _, validDimension := range validDimensions {
+			if dimension == validDimension {
+				queryParamExists = true
+				queryParameters[dimension] = option[0]
+				if len(option) != 1 {
+					multivaluedQueryParameters = append(multivaluedQueryParameters, rawDimension)
+				}
+				break
+			}
+		}
+		if !queryParamExists {
+			incorrectQueryParameters = append(incorrectQueryParameters, rawDimension)
+		}
+	}
+
+	if len(incorrectQueryParameters) > 0 {
+		return nil, apierrors.ErrorIncorrectQueryParameters(incorrectQueryParameters)
+	}
+
+	if len(multivaluedQueryParameters) > 0 {
+		return nil, apierrors.ErrorMultivaluedQueryParameters(multivaluedQueryParameters)
+	}
+
+	// Determine if any dimensions have not been set in request query parameters
+	if len(queryParameters) != len(validDimensions) {
+		for _, validDimension := range validDimensions {
+			if queryParameters[validDimension] == "" {
+				missingQueryParameters = append(missingQueryParameters, validDimension)
+			}
+		}
+		return nil, apierrors.ErrorMissingQueryParameters(missingQueryParameters)
+	}
+
+	return queryParameters, nil
+}
+
+// getUserAuthToken obtains the user auth token from the contet, expected under FlorenceIdentityKet
+func getUserAuthToken(ctx context.Context) string {
+
+	if common.IsFlorenceIdentityPresent(ctx) {
+		return ctx.Value(common.FlorenceIdentityKey).(string)
+	}
+
+	return ""
+}
+
+func (api *API) getObservationList(ctx context.Context, versionDoc *dataset.Version, queryParameters map[string]string, limit, dimensionOffset int, logData log.Data) ([]models.Observation, error) {
+
+	// Build query (observation.Filter type)
+	var dimensionFilters []*observation.Dimension
+
+	// Unable to have more than one wildcard parameter per query
+	var wildcardParameter string
+
+	// Build dimension filter object to create queryObject for neo4j
+	for dimension, option := range queryParameters {
+		if option == "*" {
+			if wildcardParameter != "" {
+				return nil, errs.ErrTooManyWildcards
+			}
+
+			wildcardParameter = dimension
+			continue
+		}
+
+		dimensionFilter := &observation.Dimension{
+			Name:    dimension,
+			Options: []string{option},
+		}
+
+		dimensionFilters = append(dimensionFilters, dimensionFilter)
+	}
+
+	queryObject := observation.DimensionFilters{
+		Dimensions: dimensionFilters,
+	}
+
+	logData["query_object"] = queryObject
+
+	log.Event(ctx, "query object built to retrieve observations from db", log.INFO, logData)
+
+	csvRowReader, err := api.graphDB.StreamCSVRows(ctx, versionDoc.ID, "", &queryObject, &limit)
+	if err != nil {
+		return nil, err
+	}
+
+	headerRow, err := csvRowReader.Read()
+	if err != nil {
+		return nil, err
+	}
+
+	defer csvRowReader.Close(ctx)
+
+	headerRowReader := csv.NewReader(strings.NewReader(headerRow))
+	headerRowArray, err := headerRowReader.Read()
+	if err != nil {
+		return nil, err
+	}
```

Comment:
Can't remember if you need to close a CSV reader but if so is there a `defer headerRowReader.close()` missing here?

Reply:
I had a look to the csv library and I would say it is not required.

Basically the csv Reader struct wraps the provided `io.Reader` with extra functionality to parse CSV records (i.e. it uses the provided reader to read records, and then parses them assuming comma separated values format). It doesn't define a Close function.

But we are creating the reader as strings.NewReader, which implements io.Reader interface. I also had a look to https://golang.org/src/strings/reader.go and it doesn't need to be closed either.

I had a look to the RowReader returned by StreamCsvRows in dp-graph, and it corresponds to a BoltRowReader, which needs to close the connection as part of Close. This isn't the case with a string reader.

Reply:
To summarise, we only need to close the `csvRowReader`

Reply:
Cool thanks for checking. Always good to double check as its very easy to miss closing a resource and it can lead to nasty bugs that are hard to find.